ds3-sdk/src/main/java/com/spectralogic/ds3client/helpers/strategy/transferstrategy/AbstractTransferStrategy.java

```diff
@@ -167,24 +183,44 @@ private synchronized void maybeSetCachedException(final Throwable t) {
 
     private void transferJobParts(final Iterable<JobPart> jobParts,
                                   final CountDownLatch countDownLatch,
-                                  final AtomicInteger numBlobsTransferred) throws IOException {
+                                  final AtomicInteger numBlobsTransferred,
+                                  final Predicate transferPredicate) throws IOException
+    {
+        final AtomicInteger numJobPartsToTransfer = new AtomicInteger(Iterables.size(jobParts));
+
         for (final JobPart jobPart : jobParts) {
-            executorService.execute(() -> {
-                try {
+            if (executorService.isShutdown()) {
+                countDownLatch.countDown();
+            } else {
+                maybeSubmitJobPartTransfer(countDownLatch, numBlobsTransferred, transferPredicate, numJobPartsToTransfer, jobPart);
+            }
+        }
+    }
+
+    private void maybeSubmitJobPartTransfer(final CountDownLatch countDownLatch, final AtomicInteger numBlobsTransferred, final Predicate transferPredicate, final AtomicInteger numJobPartsToTransfer, final JobPart jobPart) {
+        executorService.submit((Callable<Void>) () -> {
+            try {
+                if ( ! transferPredicate.test()) {
+                    countDownLatch.countDown();
```

Comment:
It seems that this can be called multiple times if we have several items submitted and a cancel occurs.  It may not matter, but this means that the .transfer method on the job will return before we've finished cancelling.